src/gluonts/model/wavenet/_network.py

```diff
@@ -547,59 +586,204 @@ def blow_up(u):
+            F,
+            feat_static_cat=feat_static_cat,
+            past_target=past_target,
+            past_observed_values=past_observed_values,
+            past_time_feat=past_time_feat,
+            future_time_feat=future_time_feat,
+            scale=scale,
+            num_samples=self.num_samples,
+            sample_based_forecast=True,
         )
-        samples = self.post_transform(samples)
+        samples = self.post_transform(samples_bin_ix)
         samples = F.broadcast_mul(scale.expand_dims(axis=1), samples)
         return samples
+
+
+class WaveNetDistributionPredictor(WaveNetPredictor):
+    """
+    Runs Wavenet generation in an auto-regressive manner using caching for
+    speedup [PKC+16]_ to produce (parametric) distribution forecast.
+
+    Same arguments as `WaveNetPredictor`.
+    """
+
+    def hybrid_forward(
+        self,
+        F,
+        feat_static_cat: Tensor,
+        past_target: Tensor,
+        past_observed_values: Tensor,
+        past_time_feat: Tensor,
+        future_time_feat: Tensor,
+        scale: Tensor,
+    ) -> Tuple[Tuple[Tensor, Tensor], Tensor]:
+        """
+        Computes the forecast distribution from the trained wavenet model.
+
+        Parameters
+        ----------
+        F
+        feat_static_cat
+            Static categorical features: (batch_size, num_cat_features)
+        past_target
+            Past target: (batch_size, receptive_field)
+        past_observed_values
+            Observed value indicator for the past target: (batch_size, receptive_field)
+        past_time_feat
+            Past time features: (batch_size, num_time_features, receptive_field)
+        future_time_feat
+            Future time features: (batch_size, num_time_features, pred_length)
+        scale
+            scale of the time series: (batch_size, 1)
+
+        Returns
+        -------
+        Tuple containing `dist_args`, `loc` and `scale`.
+
+            `dist_args` is a pair of tensors containing logarithm of bin probabilities with shape
+            (batch_size, pred_length, num_bins) and bin values with shape (num_bins,).
+
+            `loc` is a Tensor with shape (batch_size, 1).
+
+            `scale` is a Tensor with shape (batch_size, 1).
+        """
+
+        bin_probs = self.predict(
+            F,
+            feat_static_cat=feat_static_cat,
+            past_target=past_target,
+            past_observed_values=past_observed_values,
+            past_time_feat=past_time_feat,
+            future_time_feat=future_time_feat,
+            scale=scale,
+            num_samples=1,
+            sample_based_forecast=False,
+        )
+
+        # get the batch shape right
+        scale = F.broadcast_add(scale, F.zeros_like(feat_static_cat))
+        bin_centers = F.broadcast_add(
+            F.array(self.bin_values), F.zeros_like(feat_static_cat)
+        )
+
+        dist_args = F.log(bin_probs), bin_centers
+        return dist_args, scale
```

Comment:
If `loc` is needed by `DistributionOutput.distribution`, you could return `loc = F.zeros_like(scale)` here

Reply:
ya.. I had it like that but removed while fixing a test. Forgot to add it back.